src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3048,10 +3050,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                 MissingParameterTypeError(fun, vparam, pt, withTupleAddendum = !issuedMissingParameterTypeError)
                 issuedMissingParameterTypeError = true
               }
-
               setError(fun)
             }
-          } else if (numVparams == 1 && pt.typeSymbol == PartialFunctionClass) { // dodge auto-tupling with the == 1
+          }
+          else doTypedFunction(fun, resProto)
+        ftree match {
+          // dodge auto-tupling with the == 1
+          case fun0: Function if numVparams == 1 && pt.typeSymbol == PartialFunctionClass && !ftree.isErroneous && argOK(fun0, pt) =>
+            // synthesizePartialFunction will recreate param
+            val fun = resetAttrs(fun0).asInstanceOf[Function]
```

Comment:
Using `resetAttrs` worries me a little, I really don't know if can be sure enough that `typed(resetAttrs(tree))` returns a valid typed tree with the same semantics as `tree`... Could we use the enclosing untyped `fun` instead?

Reply:
Also I'm not sure about type checking the body twice, if we can avoid it... Probably it's OK, e.g. warnings are de-duplicated by position. WWDD?

Reply:
I had the same question and reservations, but I'll have to recall my conclusions to document them. The block is arbitrarily large. This was simpler than fixing up the tree, but I see you're saying just discard the typed tree and start over. I've already forgotten how to make a pandemic shruggie. I'll try it out.

Reply:
<aybe there's a way to avoid typing the `Function`'s body since we're "only" interested in the parameter type? Or maybe use `changeOwner` to re-use the typed body? Just dropping ideas.

Reply:
I think the test shows improved [?] diagnostics from the first typecheck (compared to scala 2 and 3). But reusing means fixing parameter refs. My first dismayed thought was that it's easier to re-typecheck than introduce LOC. But there's no reason to be lazy in the faul sense.

Reply:
FWIW, dotty does these things in ExpandSAMs (the rewrite with fixing up param refs and owners). That is the refchecks megaphase. Idea: `-Vphases:expandSAMs` shows the request phase name in blinking style.

Reply:
Just passing by to mention https://github.com/scala/bug/issues/5464 - it should contain references to what kind of code cannot be type checked again.

Reply:
also FWIW, I'm re-writing without re-typechecking.

Reply:
Most of my rewrites are without retyping. In the old days, I'd just break out the Wite-Out.